packages/kit/src/runtime/server/page/fetch.js

```diff
@@ -68,6 +68,19 @@ export function create_fetch({ event, options, state, route, prerender_default,
 
 				const url = new URL(request.url);
 
+				if (!request.headers.has('origin')) {
+					request.headers.set('origin', event.url.origin);
+				}
+
+				// Remove Origin, according to https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Origin#description
+				if (
+					(request.method === 'GET' || request.method === 'HEAD') &&
+					((request.mode === 'no-cors' && url.origin !== event.url.origin) ||
+						url.origin === event.url.origin)
+				) {
+					request.headers.delete('origin');
+				}
```

Comment:
The docs say nothing about removing the header when it's already present. So I think probably the most spec-compliant way is to add it when it's not present, given the mentioned conditions:
```suggestion
				// add origin, according to https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Origin#description
				if (
					!request.headers.has('origin') &&
					request.mode !== 'no-cors' &&
					(url.origin !== event.url.origin ||
						(request.method !== 'GET' && request.method !== 'HEAD'))
				) {
					request.headers.set('origin', event.url.origin);
				}
```

Reply:
@dummdidumm My understanding was, that the goal is to replicate browser behaviour. The browser does in fact remove the "Origin" for the request where no `Origin` should be present. You can easily try this yourself by doing a same-origin GET request and adding the `Origin` header

Reply:
Oh interesting ... then forget that comment. The logic still seems a little off to me, `no-cors` on a POST would not remove the origin, but it should. I think the logic can be simplified to just doing "if `no-cors` present, remove origin" (since it's a non-simple header according to the spec).
```suggestion
				if (!request.headers.has('origin')) {
					request.headers.set('origin', event.url.origin);
				}

				// Remove Origin, according to https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Origin#description
				if (
					request.mode === 'no-cors' ||
					((request.method === 'GET' || request.method === 'HEAD') &&
						url.origin === event.url.origin)
				) {
					request.headers.delete('origin');
				}
```

Reply:
@dummdidumm That is not what the description on MDN says. Also, if you do a `no-cors` POST request, the browser will add the Origin header (for both same and cross-origin requests). You can easily test that with the browser console.